core/cog-gamepad-manette.c

```diff
@@ -0,0 +1,343 @@
+/*
+ * cog-gamepad-manette.c
+ * Copyright (C) 2022 Igalia S.L.
+ *
+ * Distributed under terms of the MIT license.
+ */
+
+#include "cog-gamepad.h"
+
+#include <libmanette.h>
+#include <linux/input-event-codes.h>
+#include <wpe/wpe.h>
+
+typedef struct _Provider Provider;
+typedef struct _Gamepad  Gamepad;
+
+struct _Provider {
+    struct wpe_gamepad_provider *wk_provider;
+    ManetteMonitor              *monitor;
+    GHashTable                  *devices;
+};
+
+struct _Gamepad {
+    struct wpe_gamepad *wk_gamepad;
+    ManetteDevice      *device;
+};
+
+static void *
+provider_create(struct wpe_gamepad_provider *wk_provider)
+{
+    Provider *provider = g_new0(Provider, 1);
+
+    provider->wk_provider = wk_provider;
+    provider->devices = g_hash_table_new(NULL, NULL);
+
+    return provider;
+}
+
+static void
+provider_destroy(void *data)
+{
+    Provider *provider = data;
+
+    if (!provider)
+        return;
+
+    g_clear_object(&provider->monitor);
+    g_clear_pointer(&provider->devices, g_hash_table_unref);
+    g_free(provider);
+}
+
+static void
+add_device(Provider *provider, ManetteDevice *device)
+{
+    unsigned gamepad_id;
+
+    g_debug("gamepad: %s - connected", manette_device_get_name(device));
+
+    /* XXX(victor): this hack is for certain architectures, such as aarch64,
+     * where the memory address is bigger than unsigned int */
+    gamepad_id = GPOINTER_TO_UINT(device);
```

Comment:
that would break the libwpe API, since the gamepad identifiers  are set as uints. I think that a data type which size is arch dependent would be error prone.

Reply:
We already have that issue, becaus `unsigned int` is architecture-dependant. Maybe we should switch here to `uint64_t`. Anyway, it does not matter because we don't need things like a 32-bit build of libwpe talking to a 64-bit one of WebKit or Cog. At any rate, for the sake of convenience, we should choose a type that can fit a pointer in most (if not all) architectures, and if we don't want to use `uintptr_t`, then `size_t` is the next best choice. WDYT?

Reply:
fixed. It needs libwpe api change in https://github.com/WebPlatformForEmbedded/libwpe/pull/118